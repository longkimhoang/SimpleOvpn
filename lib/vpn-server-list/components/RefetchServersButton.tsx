import React, {useCallback, useRef} from 'react';
import {Appbar} from 'react-native-paper';
import {VpnServerRepository} from '../../vpngate-client';

export type RefetchServersButtonProps = Pick<
  VpnServerRepository,
  'fetchServers'
>;

function RefetchServersButton({fetchServers}: RefetchServersButtonProps) {
  const ref = useRef<AbortController>();

  const refetchServers = useCallback(() => {
    if (ref.current) {
      ref.current.abort();
    }

    ref.current = new AbortController();

    fetchServers({signal: ref.current.signal});
  }, [fetchServers]);

  return <Appbar.Action icon="reload" onPress={refetchServers} />;
}

export default RefetchServersButton;
