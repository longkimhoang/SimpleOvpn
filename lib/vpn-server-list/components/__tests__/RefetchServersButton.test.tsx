import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import RefetchServersButton from '../RefetchServersButton';

describe('RefetchServersButton', () => {
  test('should render correctly', () => {
    const mockFetchServers = jest.fn();

    const tree = render(
      <RefetchServersButton fetchServers={mockFetchServers} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('when tapped, should perform servers refetch', () => {
    const mockFetchServers = jest.fn();

    const {queryByTestId} = render(
      <RefetchServersButton fetchServers={mockFetchServers} />,
    );

    const buttonElement = queryByTestId(
      'vpn-server-list.button.refetch-servers',
    );

    fireEvent(buttonElement!, 'onPress');

    expect(mockFetchServers).toHaveBeenCalledTimes(1);
  });

  test('given a refetch in progress, when tapped again, should cancel the current one', () => {
    const mockFetchServers = jest.fn();

    const {queryByTestId} = render(
      <RefetchServersButton fetchServers={mockFetchServers} />,
    );

    const buttonElement = queryByTestId(
      'vpn-server-list.button.refetch-servers',
    );

    const abortSpy = jest.spyOn(AbortController.prototype, 'abort');

    fireEvent(buttonElement!, 'onPress');
    fireEvent(buttonElement!, 'onPress');

    expect(abortSpy).toHaveBeenCalledTimes(1);
  });
});
