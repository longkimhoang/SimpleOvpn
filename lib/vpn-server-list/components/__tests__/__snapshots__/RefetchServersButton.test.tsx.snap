// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefetchServersButton should render correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 0,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": "rgba(121, 116, 126, 1)",
          "borderRadius": 20,
          "borderWidth": 0,
          "elevation": 0,
          "height": 40,
          "margin": 6,
          "overflow": "hidden",
          "width": 40,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "alignItems": "center",
              "flexGrow": 1,
              "justifyContent": "center",
            },
          ]
        }
        testID="vpn-server-list.button.refetch-servers"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(73, 69, 79, 1)",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ó°‘“
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
